packages/core/echo/echo-schema/src/effect/echo-object-class.ts

```diff
@@ -0,0 +1,84 @@
+//
+// Copyright 2024 DXOS.org
+//
+
+import * as AST from '@effect/schema/AST';
+import * as S from '@effect/schema/Schema';
+import type { SimplifyMutable, Struct } from '@effect/schema/Schema';
+
+import {
+  type EchoObjectAnnotation,
+  EchoObjectAnnotationId,
+  getEchoObjectAnnotation,
+  getSchema,
+  getTypeReference,
+} from './reactive';
+
+type EchoClassOptions = {
+  partial?: true;
+};
+
+export interface EchoObjectClassType<T> {
+  new (name: Omit<T, 'id'>): T;
+}
+
+export interface EchoSchemaClass<Self, Fields> extends S.Schema<Fields> {
+  new (): Fields;
+
+  typename(): string;
+
+  isInstance(obj: unknown): obj is Self;
+}
+
+export const EchoObjectSchema = <Klass>(args: EchoObjectAnnotation) => {
+  return <
+    Options extends EchoClassOptions,
+    SchemaFields extends Struct.Fields,
+    SimplifiedFields = Options['partial'] extends boolean
+      ? SimplifyMutable<Partial<Struct.Type<SchemaFields>>>
+      : SimplifyMutable<Struct.Type<SchemaFields>>,
+    Fields = SimplifiedFields & { id: string },
+  >(
+    fields: SchemaFields,
+    options?: Options,
+  ): EchoSchemaClass<Klass, Fields> => {
+    const fieldsSchema = S.mutable(options?.partial ? S.partial(S.struct(fields)) : S.struct(fields));
+    const typeSchema = S.extend(fieldsSchema, S.struct({ id: S.string }));
+    const annotatedSchema = S.make(
+      AST.annotations(typeSchema.ast, {
+        [EchoObjectAnnotationId]: {
+          typename: args.typename,
+          version: args.version,
+        },
+      }),
+    );
+    const klass: any = class {
+      constructor() {
+        throw new Error('use E.object(MyClass, fields) to instantiate an object');
+      }
+
+      static typename() {
```

Comment:
what do you think about converting this to a property instead so its just `Schema.typename` to get the typename?